Improved error messages from clingo

becker33 opened the pull request:
Generating legible error messages from the new clingo concretizer has proven challenging, in part because the unsatisfiable cores returned from clingo solves are not minimal and cannot be minimized with reasonable performance so far.

This PR adds error message sentinels to the solve, attached to the rules that could fail a solve. The unsat core is then restricted to these messages, which makes the minimization problem tractable. Errors that can only be generated by a bug in the logic program or generating code are prefaced with `Internal error` to make clear to users that something has gone wrong on the Spack side of things.

Examples:
```
$ spack spec hdf5 ^mpich ^openmpi
Input spec
--------------------------------
hdf5
    ^mpich
    ^openmpi

Concretized
--------------------------------
==> Error: hdf5 ^mpich ^openmpi is unsatisfiable, conflicts are:
  error("All dependencies must be reachable from root")
  error("Virtual packages must be satisfied by a unique provider")
```

On an x86_64 system
```
$ spack spec zlib target=power9
Input spec
--------------------------------
zlib arch=darwin-None-power9

Concretized
--------------------------------
==> Error: zlib arch=darwin-None-power9 is unsatisfiable, conflicts are:
  error("Each node must have exactly one compiler")
  error("No satisfying compiler available is compatible with a satisfying target")
```

becker33 committed:
wip

becker33 committed:
minimize core manually

becker33 committed:
only errors messages are choices/assumptions for performance

becker33 committed:
pre-check for unreachable nodes

becker33 committed:
fix error message for multiple providers

becker33 committed:
improve python code handling errors from clingo concretizer

becker33 committed:
clean up error messages for compiler/target compiler/os mismatches

becker33 committed:
update tests for new error message

becker33 committed:
make clingo concretization errors show up in cdash reports fully

becker33 committed:
required values for old concretizer can be false-ish, check against None

becker33 committed:
fix handling of errors with virtual packages

becker33 committed:
remove duplicate code for error addressed during development

tldahlgren commented on lib/spack/spack/error.py:
```diff
@@ -117,11 +117,24 @@ class SpecError(SpackError):
+    def __init__(self, provided, required=None, constraint_type=None, conflicts=None):
+        # required is only set by the original concretizer.
+        # clingo concretizer handles error messages differently.
+        if required is not None:
```
Is it sufficient to just use `if required:` here?

becker33 replied:
No -- the required value could be `False`.

tldahlgren commented on lib/spack/spack/error.py:
```diff
@@ -117,11 +117,24 @@ class SpecError(SpackError):
+        # required is only set by the original concretizer.
+        # clingo concretizer handles error messages differently.
+        if required is not None:
+            assert not conflicts  # can't mix formats
```
Per Todd, we should be raising exceptions outside of test modules.  Should this be raising a different SpackError exception?

tgamblin replied:
If this happens it's a programming error. i.e., you should change your code to avoid it, not handle it in an exception block.  So an assertion is fine.

That said, this should have documentation for the parameters, and the split parameter thing going on here is kind of awkward.  I think it is ok in this instance because it will go away when the old concretizer does.

tldahlgren commented on lib/spack/spack/solver/asp.py:
```diff
@@ -218,11 +221,49 @@ def __init__(self, specs, asp=None):
-                "The following constraints are unsatisfiable:",
-                *sorted(str(symbol) for symbol in core))
+    def format_core(self, core):
+        assert self.control
```
Same assert issue mentioned for `error.py`.

tldahlgren commented on lib/spack/spack/solver/asp.py:
```diff
@@ -218,11 +221,49 @@ def __init__(self, specs, asp=None):
+        return sorted(str(symbol) for symbol in core_symbols)
+
+    def minimize_core(self, core):
+        assert self.control
```
Replace with SpackError exception.

tldahlgren commented on lib/spack/spack/concretize.py:
```diff
@@ -730,8 +730,8 @@ def _concretize_specs_together_new(*abstract_specs, **kwargs):
+        conflicts = result.format_minimal_cores()
+        tty.die("Unsatisfiable spec generated the following core(s):", *conflicts)
 
     return [s.copy() for s in result.specs]
```
Given the duplicated code, should this be refactored into a common method used by `cmd/solve.py` (above changes) and `concretize.py`?

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -456,6 +489,39 @@ def __init__(self):
+        ]
+        for message in error_messages:
+            self.gen.fact(fn.error(message), assumption=True)
+
```
So the unfortunate thing about this is that it's error prone: we have to write these strings in two places.  We can get around this by inspecting the `clingo` AST through the API -- this code got all the error strings for me:

```python
        from clingo.ast import ASTType, ProgramBuilder, parse_string
        with ProgramBuilder(self.control) as bld:
            def visit(node):
                if node.ast_type == ASTType.Rule:
                    for term in node.body:
                        if term.ast_type == ASTType.Literal:
                            if term.atom.ast_type == ASTType.SymbolicAtom:
                                if term.atom.symbol.name == "error":
                                    print(term.atom.symbol.arguments[0])
            path = os.path.join(parent_dir, 'concretize.lp')
            parse_string(open(path).read(), visit)
```

I put that right before the place we load `concretize.lp` in `PyclingoDriver.solve()`.  Can you test that with both the bootstrapped `clingo` and a build of latest `clingo`?  I guess I've already done the latter.

If you do that you should be able to generate your facts (or rather assumptions) right after loading `concretize.lp`.  I think to do this inspection at the same time as loading (i.e., not load twice) we would need to add AST components ourselves as part of traversal, but we can skip that -- I think sticking this in a method on `PyclingoDriver` to get the errors separately from loading the file is fine right now since parsing `concretize.lp` takes no time.

I think this will make it more natural to add more error messages to `concretize.lp` -- we can just add more `error()` lines to rules.  If we need to add messages to more than just rules I think we can worry about that in a follow-on PR.

I think `PyclingoDriver` is the right place for this vs. setup b/c it's really part of the logic program -- not part of the input facts (even if `error()` assumptions are added in a similar way).

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -218,11 +221,49 @@ def __init__(self, specs, asp=None):
+
+        return sorted(str(symbol) for symbol in core_symbols)
+
+    def minimize_core(self, core):
```
This needs a good docstring.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -218,11 +221,49 @@ def __init__(self, specs, asp=None):
-            tty.msg(
-                "The following constraints are unsatisfiable:",
-                *sorted(str(symbol) for symbol in core))
+    def format_core(self, core):
```
This needs a good docstring

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -218,11 +221,49 @@ def __init__(self, specs, asp=None):
+        return min_core
+
+    def minimal_cores(self):
+        return [self.minimize_core(core) for core in self.cores]
```
Do you ever see more than one?  I never have but it doesn't mean it couldn't happen.

becker33 replied:
You can see more than one, but I don't remember what I did that made that happen. I haven't seen it with the most recent version of the code though, so I'm not certain whether it was an artifact of an earlier error.

tgamblin replied:
I think it’s just data-dependent. I just hadn’t seen it at all. Cool!

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1442,6 +1516,9 @@ def setup(self, driver, specs, tests=False):
         self.define_target_constraints()
 
+        self.gen.h1("Rule explanations")
+        self.error_messages()
```
Delete with move of error setup to `PyclingoDriver`

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -311,15 +352,16 @@ def h2(self, name):
         self.out.write('\n')
 
-    def fact(self, head):
+    def fact(self, head, assumption=False):
```
update docstring with arg descriptions

tgamblin commented on lib/spack/spack/concretize.py:
```diff
@@ -730,8 +730,8 @@ def _concretize_specs_together_new(*abstract_specs, **kwargs):
-        result.print_cores()
-        tty.die("Unsatisfiable spec.")
+        conflicts = result.format_minimal_cores()
+        tty.die("Unsatisfiable spec generated the following core(s):", *conflicts)
```
I think we should print out something like:

```python
tty.error("Unsatisfiable spec.", "Analyzing conflicts. This may take some time...")
conflicts = result.format_minimal_cores()
tty.die("Concretization failed with the following issues:", *conflicts)
```

So that you know before the core minimization that it's going to happen (you could also Ctrl-C at this point if you know what the problem likely is).

becker33 replied:
With only the error messages in the core, core minimization is only taking <2 seconds for solves that take >10 seconds. I'm not sure how useful that message is given the relative runtimes.

tgamblin replied:
I think it's still useful *in case* it starts taking more time.

becker33 committed:
refactor unsat core formatting

becker33 committed:
update docstring for fact generation

becker33 committed:
read error messages from clingo rather than duplicating in asp.py

tldahlgren reviewed (commented):
Thanks for the refactor.

becker33 committed:
use clingo.ast.parse_files for backwards compatibility

tgamblin committed:
clingo: make import of clingo.ast parsing routines robust to clingo version

Older `clingo` has `parse_string`; newer `clingo` has `parse_files`.  Make the
code work wtih both.

haampie commented:
parse_files is top-level clingo.parse_files, but even with that change tests do not pass (    'clingo.ast.AST' object has no attribute 'ast_type')

tgamblin committed:
Greg was right; it's nicer if we just print the error once.

tgamblin committed:
make version error message more general (it's not just internal)

tgamblin committed:
make AST access functions backward-compatible with clingo 5.4.0

Clingo AST API has changed since 5.4.0; make some functions to help us
handle both versions of the AST.

tgamblin reviewed (approved):
@becker33 this is ready IMO. Can you squash it?
